README.md

```diff
@@ -1,27 +1,508 @@
+
+## `setState` in detail
+
+Now that we built our blog, let's have a look at something something a bit more
+interesting: A modal.
+
+![Modal](./diagrams/modal.png)
+
+Let's first start off laying out our component structure:
+
+```js
+class App extends React.Component {
+    render () {
+        return (
+            <div>
+                <Modal />
+            </div>
+        )
+    }
+}
+```
+
+A modal can either be open or closed, so our application needs to keep track of
+that. The `<Modal />` component itself only _knows_ whether or not it is open,
+it can't "open" itself — instead it has to expose some API for other components
+for opening it.
+
+Therefore we need to keep track of whether or not the modal is open in its
+parent component. In this case, that's the `<App />` component. On page load,
+the modal shouldn't be open. The initial state of the modal is closed
+(`isModalOpen = false`). We can initialize a component's state in its
+constructor:
+
+```js
+class App extends React.Component {
+    constructor (props) {
+        super(props)
+        this.state = {
+            // modal should be closed on page load
+            isModalOpen: false
+        }
+    }
+    // ...
+}
+```
+
+We then need to "tell" the `<Modal />` component whether or not it is open. The
+only way to communicate with the component is via its `props`. In this case,
+we would add a new `isOpen` prop to the `Modal` component:
+
+```js
+class App extends React.Component {
+    // ...
+    render () {
+        return (
+            <div>
+                <Modal isOpen={this.state.isModalOpen} />
+            </div>
+        )
+    }
+}
+```
+
+Obviously our users are going to be very disappointed if they aren't able to
+open the popup, so let's add a separate button that triggers the modal:
+
+```js
+class App extends React.Component {
+    // ...
+    openModal () {
+        this.setState({ isModalOpen: true })
+    }
+    render () {
+        return (
+            <div>
+                <button onClick={this.openModal}>Open the modal!</button>
+                <Modal isOpen={this.state.isModalOpen} />
+            </div>
+        )
+    }
+}
+```
+
+Now if we click the `Open the modal!` button, we update the `<App />`
+component's internal state, thus enforcing an update on the `<App />` component.
+
+Since `this.state.isModalOpen` evaluates to `true` now, the modal is being
+opened:
+
+```js
+    render () {
+        return (
+            <div>
+                <button onClick={this.openModal}>Open the modal!</button>
+                <Modal isOpen={true} />
+            </div>
+        )
+    }
+```
+
+But what does our modal component actually look like? Up until now, we didn't
+talk about the `<Modal />` component's render method.
+
+As we just saw, the modal accepts an `isOpen` prop, which it then uses in order
+to add an `modal--is-open` class to its container:
+
+```js
+const Modal = ({ isOpen }) =>
```

Comment:
Can we use a regular ES2015 React component for this instead of a pure component?